lean4-infoview/src/infoview/goalLocation.tsx

```diff
@@ -76,66 +76,125 @@ export interface Locations {
+                return
+            }
+            ;(e as any)._DetectHoverSpanSeen = {}
+            if (!b) {
+                setHoverState('off')
+            } else if (e.ctrlKey || e.metaKey) {
+                setHoverState('ctrlOver')
+            } else {
+                setHoverState('over')
+            }
+        }
+    }
+
+    const onPointerMove = (e: React.PointerEvent<HTMLSpanElement>) => {
+        if (e.ctrlKey || e.metaKey) {
+            setHoverState(st => (st === 'over' ? 'ctrlOver' : st))
+        } else {
+            setHoverState(st => (st === 'ctrlOver' ? 'over' : st))
+        }
+    }
+
+    const onKeyDown = (e: React.KeyboardEvent<HTMLSpanElement>) => {
+        if (e.key === 'Control' || e.key === 'Meta') {
+            setHoverState(st => (st === 'over' ? 'ctrlOver' : st))
+        }
+    }
+
+    const onKeyUp = (e: React.KeyboardEvent<HTMLSpanElement>) => {
+        if (e.key === 'Control' || e.key === 'Meta') {
+            setHoverState(st => (st === 'ctrlOver' ? 'over' : st))
+        }
     }
 
-    const innerSpanClassName: string =
-        'highlightable ' + (locs && loc && locs.isSelected(loc) ? 'highlight-selected ' : '')
-
-    const setHoverStateAll: React.Dispatch<React.SetStateAction<HoverState>> = React.useCallback(
-        val => {
-            setHoverState(val)
-            if (setParentHoverState) setParentHoverState(val)
-        },
-        [setParentHoverState],
-    )
-
-    return (
-        <DetectHoverSpan
-            {...props}
-            setHoverState={setHoverStateAll}
-            className={spanClassName}
-            onClick={e => {
-                // On shift-click, if we are in a context where selecting subexpressions makes sense,
-                // (un)select the current subexpression.
-                if (e.shiftKey && locs && loc) {
-                    locs.setSelected(loc, on => !on)
-                    e.stopPropagation()
-                }
-                if (props.onClick) props.onClick(e)
-            }}
-            onPointerDown={e => {
-                // Since shift-click on this component is a custom selection, when shift is held prevent
-                // the default action which on text is to start a text selection.
-                if (e.shiftKey) e.preventDefault()
-                if (props.onPointerDown) props.onPointerDown(e)
-            }}
-        >
-            {/* Note: we use two spans so that the two `highlight`s don't interfere. */}
-            <span className={innerSpanClassName}>{props.children}</span>
-        </DetectHoverSpan>
-    )
+    const onClick = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
+        // On shift-click, if we are in a context where selecting subexpressions makes sense,
+        // (un)select the current subexpression.
+        if (selectionSettings.highlightOnSelection && locs && e.shiftKey) {
+            locs.setSelected(selectionSettings.loc, on => !on)
```

Comment:
This is because `setSelected` changes the `LocationsContext`, which is in turn read in every `InteractiveCodeTag` and hence triggers a global re-render. There's probably some way to fix this by refactoring the whole `LocationsContext` control flow, but I don't think it's easy.

This takes about 400ms on this machine, which isn't great, but seems to be acceptable. Do you think that fixing this is important now?

Reply:
I don't this it's urgent. Thanks for looking into it.